relay-server/src/actors/project_upstream.rs

```diff
@@ -191,191 +230,240 @@ impl UpstreamProjectSource {
-                for channel in channels_batch.values_mut() {
-                    channel.attempts += 1;
-                }
-                relay_log::debug!("sending request of size {}", channels_batch.len());
-                metric!(
-                    histogram(RelayHistograms::ProjectStateRequestBatchSize) =
-                        channels_batch.len() as u64
-                );
-
-                let query = GetProjectStates {
-                    public_keys: channels_batch.keys().copied().collect(),
-                    full_config: self.config.processing_enabled(),
-                    no_cache: channels_batch.values().any(|c| c.no_cache),
-                };
-
-                // count number of http requests for project states
-                metric!(counter(RelayCounters::ProjectStateRequest) += 1);
-
-                UpstreamRelay::from_registry()
-                    .send(SendQuery(query))
-                    .map_err(|_| ProjectError::ScheduleFailed)
-                    .map(move |response| (channels_batch, response))
-            })
-            .collect();
+    /// Executes an upstream request to fetch project configs.
+    ///
+    /// This assumes that currently no request is running. If the upstream request fails or new
+    /// channels are pushed in the meanwhile, this will reschedule automatically.
+    async fn fetch_states(
+        config: Arc<Config>,
+        channels: ChannelsBatch,
+        inner_tx: mpsc::UnboundedSender<Vec<UpstreamResponse>>,
+    ) {
+        let request_start = Instant::now();
+        let batch_size = config.query_batch_size();
+        let cache_batches = channels.cache_channels.into_iter().chunks(batch_size);
+        let nocache_batches = channels.nocache_channels.into_iter().chunks(batch_size);
+
+        let mut requests = vec![];
+        for channels_batch in cache_batches.into_iter().chain(nocache_batches.into_iter()) {
+            let mut channels_batch: ProjectStateChannels = channels_batch.collect();
+            for channel in channels_batch.values_mut() {
+                channel.attempts += 1;
+            }
+            relay_log::debug!("sending request of size {}", channels_batch.len());
+            metric!(
+                histogram(RelayHistograms::ProjectStateRequestBatchSize) =
+                    channels_batch.len() as u64
+            );
+
+            let query = GetProjectStates {
+                public_keys: channels_batch.keys().copied().collect(),
+                full_config: config.processing_enabled(),
+                no_cache: channels_batch.values().any(|c| c.no_cache),
+            };
+
+            // count number of http requests for project states
+            metric!(counter(RelayCounters::ProjectStateRequest) += 1);
+
+            let future_request = compat::send(UpstreamRelay::from_registry(), SendQuery(query))
+                .map(|response| match response {
```

Comment:
_nit_: You can wrap the asynchronous parts of the loop into an `async` block to avoid using the future trait extensions (might need to be async move):

```rust
// ...

for channels_batch in cache_batches {
  // ...
  let query = GetProjectStates { .. };
  
  requests.push(async move {
    let request = compat::send(...);
    match request.await {
      Ok(response) => UpstreamResponse { .. },
      Err(_) => UpstreamResponse { .. },
    }
  });
}

let responses = future::join_all(requests).await;
// ...
```